lib/file-source/src/buffer.rs

```diff
@@ -0,0 +1,231 @@
+            // FileWatcher.
+            return Ok(None);
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::read_until_with_max_size;
+    use bytes::{BufMut, BytesMut};
+    use quickcheck::{QuickCheck, TestResult};
+    use std::io::Cursor;
+    use std::num::NonZeroU8;
+    use std::ops::Range;
+
+    fn qc_inner(chunks: Vec<Vec<u8>>, delim: u8, max_size: NonZeroU8) -> TestResult {
+        // The `global_data` is the view of `chunks` as a single contiguous
+        // block of memory. Where `chunks` simulates what happens when bytes are
+        // fitfully available `global_data` is the view of all chunks assembled
+        // after every byte is available.
+        let mut global_data = BytesMut::new();
+
+        // `DelimDetails` describes the nature of each delimiter found in the
+        // `chunks`.
+        #[derive(Clone)]
+        struct DelimDetails {
+            /// Index in `global_data`, absolute offset
+            global_index: usize,
+            /// Index in each `chunk`, relative offset
+            interior_index: usize,
+            /// Whether this delimiter was within `max_size` of its previous
+            /// peer
+            within_max_size: bool,
+            /// Which chunk in `chunks` this delimiter was found in
+            chunk_index: usize,
+            /// The range of bytes that this delimiter bounds with its previous
+            /// peer
+            byte_range: Range<usize>,
+        }
+
+        // Move through the `chunks` and discover at what positions an instance
+        // of `delim` exists in the chunk stream and whether that `delim` is
+        // more than `max_size` bytes away from the previous `delim`. This loop
+        // constructs the `facts` vector that holds `DelimDetails` instances and
+        // also populates `global_data`.
+        let mut facts: Vec<DelimDetails> = Vec::new();
+        let mut global_index: usize = 0;
+        let mut previous_offset: usize = 0;
+        for (i, chunk) in chunks.iter().enumerate() {
+            for (interior_index, byte) in chunk.iter().enumerate() {
+                global_data.put_u8(*byte);
+                if *byte == delim {
+                    let span = global_index - previous_offset;
+                    let within_max_size = span <= max_size.get() as usize;
+                    facts.push(DelimDetails {
+                        global_index,
+                        within_max_size,
+                        chunk_index: i,
+                        interior_index,
+                        byte_range: (previous_offset..global_index),
+                    });
+                    previous_offset = global_index + 1;
+                }
+                global_index += 1;
+            }
+        }
+
+        // Our model is only concerned with the first valid delimiter in the
+        // chunk stream. As such, discover which that it and record it
```

Comment:
```suggestion
        // chunk stream. As such, discover which that is and record it
```

Reply:
Hmm, "which that" is an awkward construction in itself. Good call-out.